### PR TITLE
Copy container parser from jxl-oxide

### PR DESCRIPTION
**Pull request by tirr-c**
Almost the exact copy of jxl-oxide, except the error type. I'd like to add some notice about original code somewhere (jxl-oxide is MIT licensed after all), but I'm not entirely sure about where to add those...

### CODE REVIEW

File: jxl/src/container/mod.rs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+                    if buf.starts_with(&Self::CONTAINER_SIG) {
+                        // tracing::debug!("Container signature found");
+                        *state = DetectState::WaitingBoxHeader;
+                        buf.drain(..Self::CONTAINER_SIG.len());
+                        continue;
+                    }
+                    if !Self::CODESTREAM_SIG.starts_with(buf)
+                        && !Self::CONTAINER_SIG.starts_with(buf)
+                    {
+                        // tracing::error!("Invalid signature");
```
Comment (veluca93) [inline jxl/src/container/mod.rs:107]: Feel free to uncomment those, but I would only use DEBUG/TRACE levels so they are omitted by default in builds.
(I believe there is no way to set the tracing static log level per-crate)

File: jxl/src/container/mod.rs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+                            kind: BitstreamKind::Invalid,
+                            bytes_left: None,
+                        };
+                        continue;
+                    }
+                    return Ok(());
+                }
+                DetectState::WaitingBoxHeader => match ContainerBoxHeader::parse(buf)? {
+                    HeaderParseResult::Done { header, size } => {
+                        buf.drain(..size);
```
Comment (veluca93) [inline jxl/src/container/mod.rs:118]: I believe this is potentially quadratic if many small boxes are present and a large input is fed to the box decoder.
In general, unconditionally buffering the received input is best avoided too.

File: jxl/src/container/mod.rs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+            | DetectState::WaitingJxlpIndex(..)
+            | DetectState::InAuxBox { .. } => BitstreamKind::Container,
+            DetectState::InCodestream { kind, .. } | DetectState::Done(kind) => kind,
+        }
+    }
+
+    pub fn feed_bytes(&mut self, input: &[u8]) -> Result<(), Error> {
+        let state = &mut self.state;
+        let buf = &mut self.buf;
+        buf.extend_from_slice(input);
```
Comment (veluca93) [inline jxl/src/container/mod.rs:85]: Here and below, I would use buf.try_reserve() to make sure that we do not get panics due to failed allocations

File: jxl/src/container/mod.rs (hunk 2 above)
Comment (tirr-c) [reply jxl/src/container/mod.rs]: Yeah, I just realized while thinking about how to reduce buffering. Addressed in c3e8a51.

File: jxl/src/container/mod.rs (hunk 3 above)
Comment (tirr-c) [reply jxl/src/container/mod.rs]: Also addressed in c3e8a51.

File: jxl/src/container/box_header.rs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        header_size: usize,
+    },
+    NeedMoreData,
+}
+
+impl ContainerBoxHeader {
+    pub(super) fn parse(reader: &super::ConcatSlice<'_, '_>) -> Result<HeaderParseResult, Error> {
+        let mut buf = [0u8; 16];
+        let buf = reader.peek(&mut buf);
+        if buf.len() < 8 {
```
Comment (veluca93) [inline jxl/src/container/box_header.rs:30]: Doesn't this follow anyway from the default case in the match?

**Review by veluca93 — APPROVED** [review]

File: jxl/src/container/mod.rs [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Default)]
+enum JxlpIndexState {
+    #[default]
+    Initial,
+    SingleJxlc,
+    Jxlp(u32),
+    JxlpFinished,
+}
+
+struct ConcatSlice<'first, 'second> {
```
Comment (veluca93) [inline jxl/src/container/mod.rs:73]: Maybe it makes sense to move this in util/?

File: jxl/src/container/box_header.rs (hunk 4 above)
Comment (tirr-c) [reply jxl/src/container/box_header.rs]: Ooh, right -- I've removed the if statement.

File: jxl/src/container/mod.rs (hunk 1 above)
Comment (veluca93) [reply jxl/src/container/mod.rs]: https://users.rust-lang.org/t/turn-off-tracing-at-compile-time-for-specific-crate/104337/2

actually, maybe this is a better way... (feel free to do it in a separate PR)

**Review by veluca93 — APPROVED** [review]

File: jxl/src/container/mod.rs (hunk 1 above)
Comment (tirr-c) [reply jxl/src/container/mod.rs]: I'd prefer doing that in a separate PR.

Could you merge this if it looks good to you? The branch protection rule doesn't seem to allow me to merge PRs...

**Pull request merged**